fe/fe-core/src/main/java/org/apache/doris/analysis/Expr.java

```diff
@@ -871,12 +871,20 @@ public String toSql() {
         return (printSqlInParens) ? "(" + toSqlImpl() + ")" : toSqlImpl();
     }
 
+    public String toDigest() {
+        return (printSqlInParens) ? "(" + toDigestImpl() + ")" : toDigestImpl();
+    }
+
     /**
      * Returns a SQL string representing this expr. Subclasses should override this method
      * instead of toSql() to ensure that parenthesis are properly added around the toSql().
      */
     protected abstract String toSqlImpl();
 
+    protected String toDigestImpl() {
+        return toSqlImpl();
```

Comment:
Maybe we should throw `NotImplementedException` here?
So that we can ensure that every newly added Expr subclass will override this method?

Reply:
Thx for your reply.
The abstract method `toSqlImpl()` will be overrided in each subclass. So there is no need to throw out exceptions.
The only situation is that a newly added subclass, which its sql digest needs to be different from sql impl, should override this method.

Added comments over this method to remind developers.